APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+
+    matrix = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
+
+The mitigation here is the exception policy:  In this case the author could
+surround the matrix with the ``#fmt: off``/``#fmt: on`` escape hatch, as this
+clearly makes the code more readable.  While many examples are less clear than
+the above, and indeed in most cases "readability" is a subjective concept, this
+mechanism allows for such cases to be explicitly excepted where the developer
+thinks the improved readability warrants it.
+
+History
+~~~~~~~
+This changes almost every Python file. Git blame can skip the commit, but git
+diff won't. But we are already doing other bulk changes that hit every file.
+
+Code correctness
+~~~~~~~~~~~~~~~~
+Black guarantees semantic equivalence of the formatted code, and will not (`except in a few limited cases <https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#ast-before-and-after-formatting>`_) change the code AST.
+
+Migration effort
+~~~~~~~~~~~~~~~~
+Enthusiastic champions of the effort will largely do the work, but we recognize
+and appreciate that additional effort will be required from maintainers that are
+neutral or against the adoption of Black.
+
+Backports / release branches
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This is addressed in the Implementation section.
+
+New contributors
+~~~~~~~~~~~~~~~~
+New contributors may be more discouraged by more steps. However:
+
+This *replaces* the existing PEP8 check, so that it is not "another step" To
+make it better, leave very clear explicit messages in the status checks on GH
+(not putting up a barrier) Scikit-learn had a few issues with contributors but
+not at a level that caused regret about the decision to adopt black formatting.
+One developer described the change as a "net benefit".
+
+Removing the humanity
+~~~~~~~~~~~~~~~~~~~~~
+Some developers may feel that individual style and humanity gets lost in
+autoformatting. We hope to convey that humanity and individuality is expressed
+more fully in the elegance and clarity of the code logic and implementation. In
+most cases the placement of white space and line breaks can be viewed as a
+mechanistic process that humans need not control.
+
+Exception Policy
+^^^^^^^^^^^^^^^^
+As described above, there are cases Black formatting makes the code
+substantially more difficult to read or understand.. These should be treated
+explicitly as exceptions, using the ``#fmt: off``/``#fmt: on`` escape hatch for
+multi-line exceptions, or ``#fmt: skip`` for single-line exceptions. **This is
+allowed if the code author and PR reviewer(s) agree that the code line or block
+should be excluded from Black formatting.**
+
+
+Branches and pull requests
+--------------------------
+
+- `Add configuration for black autoformatter <https://github.com/astropy/astropy/pull/13253>`_
+- `Apply black to modeling <https://github.com/astropy/astropy/pull/13254>`_
+
+Implementation
```

Comment:
This section is missing something about how we will deal with release branches, and when the change will be done, both to minimize impact on release branches and open PRs. As far as I can see there are two options:

* Make the change just before the first 6.0 release candidate is made, because at that point the release branches will no longer be active and open PRs may be at a minimum.
* Have a sprint some other time to minimize open PRs, then pick an arbitrary date to make the change and simultaneously reformat ``main`` and any active release branch.

Reply:
I like the second option. 

The main concern there is the extent of exceptions that are put in by hand for `main`. The APE is explicit in allowing exceptions, but I don't see an easy way to apply those exceptions to formatting any active release branch -- they would be formatted with a single run of `black` on the entire repo. I'm hoping that in practice the exceptions will be rare and won't generate conflicts.

Reply:
I agree as well. Indeed we may want to plan a sprint to coincide with the 6.0 release so that we can push all these changes in quickly. The creation of the initial `Black` changes will be simple, but reviewing the changes for any possible exceptions will be the tedious part, and we may want to have people ready to do this work.

I worry about any remaining open PRs after `Black` changes are merged. We should work with the contributors of any of the remaining open PRs to both warn them and help them get their PRs migrated.

Reply:
I was wondering about a hybrid of the options -- 

After the normal v5.2 feature freeze on 2022-10-28, then have a 2-week period for doing the `Black` transition of all sub-packages. I think the initial PR's could be done by a core team of a few, but this requires some buy-in from sub-package maintainers to do the reviews. By doing this in a special post-freeze period this might minimize the impact on open PR's. Many PR's that stay open through a feature freeze are not being actively worked.

The 6.0 freeze is not until 2023-10-27 so I hope we don't need to wait that long!

Reply:
I like the sound of this (esp the earlier deadline 😆 )

Reply:
Release managers @astrofrog @bsipocz @saimn @embray, and others - are you OK with the plan of a 2-week post-freeze period to do the Black transition PRs for sub-packages? Or does anyone have a better plan?

Reply:
[`darken`](https://pypi.org/project/darker/) can apply `black` formatting to only the lines that are being edited. This would allow for a transition period during which all new code would be `black`-formatted, so when the final switch to `black` finally happens the diffs would be smaller.

Reply:
I'm happy with a 2-week post-freeze period, though we will need to decide what to do for sub-packages where maintainers end up being unavailable to review.

I think we might want to not run black on the LTS branch but as that branch has existed for almost a year the number of bug fixes that backport cleanly is going to be more and more limited, so maybe we just don't format the LTS branch and deal with conflicts by hand.

Reply:
Personally, I've found ``darken`` more finicky than ``black`` to get working with IDEs. Not saying I'm against it, but going all in would be my preference.